File: robert-server-ws-rest/src/test/java/fr/gouv/stopc/robertserver/ws/test/matchers/TimeDriftMatcher.java
package fr.gouv.stopc.robertserver.ws.test.matchers;

import java.time.Duration;

import static fr.gouv.stopc.robertserver.ws.test.LogbackManager.assertThatInfoLogs;
import static fr.gouv.stopc.robertserver.ws.test.LogbackManager.assertThatWarnLogs;
import static fr.gouv.stopc.robertserver.ws.test.MongodbManager.assertThatRegistrationTimeDriftForUser;
import static org.assertj.core.api.HamcrestCondition.matching;
import static org.assertj.core.data.Offset.offset;
import static org.hamcrest.Matchers.matchesPattern;

public class TimeDriftMatcher {

    /**
     * Verifies logs about exceeded time drift are produced and the delta is stored
     * in the user registration.
     */
    public static void verifyExceededTimeDriftIsProperlyHandled(final String user, final Duration timeDrift) {
        assertThatWarnLogs().areExactly(
                1,
                matching(
                        matchesPattern(
                                "Witnessing abnormal time difference -?\\d+ between client: [0-9-T:.Z=E]+ and server: [0-9-T:.Z=E]+"
                        )
                )
        );
        assertThatInfoLogs()
                .contains("Discarding authenticated request because provided time is too far from current server time");

        assertThatRegistrationTimeDriftForUser(user)
                .isCloseTo(-timeDrift.toSeconds(), offset(2L));
    }

}

File: robert-server-ws-rest/src/test/kotlin/fr/gouv/stopc/robertserver/ws/test/matchers/TimeDriftMatcher.kt
package fr.gouv.stopc.robertserver.ws.test.matchers

import fr.gouv.stopc.robertserver.test.MongodbManager.Companion.assertThatRegistrationForIdA
import fr.gouv.stopc.robertserver.test.assertThatInfoLogs
import org.assertj.core.api.HamcrestCondition.matching
import org.assertj.core.condition.VerboseCondition
import org.assertj.core.condition.VerboseCondition.verboseCondition
import org.hamcrest.Matchers.matchesPattern
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

/**
 * Verifies logs about exceeded time drift are produced and the delta is stored
 * in the user registration.
 */
fun verifyExceededTimeDriftIsProperlyHandled(user: String, timeDrift: Duration) {
    assertThatInfoLogs().areExactly(
        1,
        matching(
            matchesPattern(
                "Auth denied on POST /api/v6/[^ ]+: Too much clock skew [^ ]+ between client \\([0-9-T:.Z=E]+\\) and server \\([0-9-T:.Z=E]+\\)"
            )
        )
    )

    assertThatRegistrationForIdA(user)
        .hasEntrySatisfying("lastTimestampDrift", timeDriftCloseTo(timeDrift))
}

fun timeDriftCloseTo(timeDrift: Duration): VerboseCondition<Any> = verboseCondition(
    { value -> timeDrift.absoluteValue - (value as Long).seconds.absoluteValue <= 2.seconds },
    "a number with absolute value close to $timeDrift +/- 2s",
    { value -> " was $value" }
)
